### ISSUE TITLE
Multi-monitor support for grids

### ISSUE THREAD

**Issue by chilimangoes** [synkarius/caster#93]
Does anyone else use multiple monitors?

Currently, it doesn't seem like any of the grid commands (e.g. Legion, Rainbow) work on anything other than the primary monitor. How difficult would it be to modify them to display on a different monitor? I was thinking something like "Rainbow 2" or "Rainbow 3" to display the rainbow grid on monitor two or three, respectively.

**Comment by Zone22** [comment]
I have two monitors and could help with testing.

**Comment by chilimangoes** [comment]
From [this script](https://github.com/dictation-toolbox/dragonfly-scripts/blob/7c31b3980dca68989869d56539d8de9eca439bc0/_winctrl.py), it appears that dragonfly has some built-in support for multiple monitors. I think all we would need to do is [enumerate dragonfly.monitors](https://github.com/dictation-toolbox/dragonfly-scripts/blob/7c31b3980dca68989869d56539d8de9eca439bc0/_winctrl.py#L103-L104) and then [get the rectangle of the selected monitor](https://github.com/dictation-toolbox/dragonfly-scripts/blob/7c31b3980dca68989869d56539d8de9eca439bc0/_winctrl.py#L304-L305).

I'm looking at grids.py and I see that the constructor for TkTransparent accepts a `dimensions` parameter, which is then exposed by the concrete grid classes, such as RainbowGrid and DouglasGrid as `grid_size`. I believe that setting this parameter to the rectangle of the selected monitor is what we need to do, but I need to do some more digging to figure out how to do that.

**Comment by chilimangoes** [comment]
I implemented multi-monitor support for the Rainbow and Douglas grids last night. Still need to tackle the Legion grid.

@synkarius Could you take a look at [this commit](https://github.com/chilimangoes/caster/commit/d4c919464d219a232f8fbbdeab3e6590ecbf2582) and let me know what you think of this approach?

**Comment by synkarius** [comment]
Looks pretty thorough to me. My only request would have been to make the monitor number optional in the spec, which you did.

Actually I'm a little surprised that you got away with importing something from Dragonfly in an external script. A lot of Dragonfly modules import other modules which import other modules which cause Dragonfly to re-initialize (and crash), but it seems `monitors` is safe.

**Comment by synkarius** [comment]
Rereading this, I realize you asked what I think of the approach. I like it.

**Comment by chilimangoes** [comment]
I meant approach in a very general sense, i.e. if you had feedback from either a design or technical standpoint. :) Your point about importing Dragonfly modules is a good one – I'll have to remember that in the future. Perhaps I should also take a look at the monitor code and consider moving it into caster. What do you think?

Another thing I just realized, I should probably replace the integer monitor numbers with an IntegerRefST for consistency.

The next couple of days are going to be pretty busy for me, so I may not get back to this until the weekend.

**Comment by synkarius** [comment]
In that case, I think the approach is solid. The `monitors` module was a good find, seems to handle monitor positioning and size, which are the obvious concerns. One odd thing I remember from when I originally made these was that if I made them the entire size of the screen, they would cover the start menu, but making them even a pixel less, the start menu would cover them. So long as the latter doesn't happen, I think we're golden.

I wouldn't worry too much about the `monitors` module unless you have some specific scenario in mind. I don't think that Dragonfly source is going to change anytime soon.

Also, I agree that IntegerRefST would be better.

**Comment by synkarius** [comment]
The grids don't cover the taskbar like they used to. At first, I was a little annoyed about that, but then I thought about new users not knowing the "exit" command, and now I'm thinking maybe it's better this way. You can't lock yourself out of your own screen anymore. And if you really need to get to the taskbar, you can always pick a position near it and then use the "curse" command.

**Comment by chilimangoes** [comment]
Hmm, that's strange since I didn't change anything about how the window is positioned once I've got the coordinates for the monitor. Considering your previous comment, maybe there's a slight difference in the coordinates between `monitors[n].rectangle` and the screen width/height returned by TkInter.

I actually got caught by the issue of not knowing the exit command and not remembering how to trigger the coordinates in the Douglas grid when I was first experimenting with caster. :) I was able to close it using Alt+F4 or "close window" IIRC, but it was a little disconcerting for a minute. It would probably be good to add "cancel/escape" as a way for exiting, like with Dragon's built-in mouse grid.

In any case, I've made some decent progress on recreating the grid window using ctypes and that will give us a lot more control over little usability issues like this. I hope to have a first cut of that working in the next couple of weeks, depending on how busy I am.
